Kernel/ACPI/MultiProcessorParser.h

```diff
@@ -0,0 +1,241 @@
+        u16 ext_table_length;
+        u8 ext_table_checksum;
+        u8 reserved;
+        EntryHeader entries[];
+    };
+
+    enum class ConfigurationTableEntryType {
+        Processor = 0,
+        Bus = 1,
+        IOAPIC = 2,
+        IO_Interrupt_Assignment = 3,
+        Local_Interrupt_Assignment = 4,
+        SystemAddressSpaceMapping = 128,
+        BusHierarchyDescriptor = 129,
+        CompatibilityBusAddressSpaceModifier = 130
+    };
+
+    enum class ConfigurationTableEntryLength {
+        Processor = 20,
+        Bus = 8,
+        IOAPIC = 8,
+        IO_Interrupt_Assignment = 8,
+        Local_Interrupt_Assignment = 8,
+        SystemAddressSpaceMapping = 20,
+        BusHierarchyDescriptor = 8,
+        CompatibilityBusAddressSpaceModifier = 8
+    };
+
+    struct [[gnu::packed]] ExtEntryHeader
+    {
+        u8 entry_type;
+        u8 entry_length;
+    };
+
+    struct [[gnu::packed]] ProcessorEntry
+    {
+        EntryHeader h;
+        u8 local_apic_id;
+        u8 local_apic_version;
+        u8 cpu_flags;
+        u32 cpu_signature;
+        u32 feature_flags;
+        u8 reserved[8];
+    };
+
+    struct [[gnu::packed]] BusEntry
+    {
+        EntryHeader h;
+        u8 bus_id;
+        char bus_type[6];
+    };
+
+    struct [[gnu::packed]] IOAPICEntry
+    {
+        EntryHeader h;
+        u8 ioapic_id;
+        u8 ioapic_version;
+        u8 ioapic_flags;
+        u32 ioapic_address;
+    };
+
+    enum class InterruptType {
+        INT = 0,
+        NMI = 1,
+        SMI = 2,
+        ExtINT = 3,
+    };
+
+    struct [[gnu::packed]] IOInterruptAssignmentEntry
+    {
+        EntryHeader h;
+        u8 interrupt_type;
+        u8 polarity;
+        u8 trigger_mode;
+        u8 source_bus_id;
+        u8 source_bus_irq;
+        u8 destination_ioapic_id;
+        u8 destination_ioapic_intin_pin;
+    };
+
+    struct [[gnu::packed]] LocalInterruptAssignmentEntry
+    {
+        EntryHeader h;
+        u8 interrupt_type;
+        u8 polarity;
+        u8 trigger_mode;
+        u8 source_bus_id;
+        u8 source_bus_irq;
+        u8 destination_lapic_id;
+        u8 destination_lapic_lintin_pin;
+    };
+
+    enum class SystemAddressType {
+        IO = 0,
+        Memory = 1,
+        Prefetch = 2,
+    };
+
+    struct [[gnu::packed]] SystemAddressSpaceMappingEntry
+    {
+        ExtEntryHeader h;
+        u8 bus_id;
+        u8 address_type;
+        u64 address_base;
+        u64 length;
+    };
+
+    struct [[gnu::packed]] BusHierarchyDescriptorEntry
+    {
+        ExtEntryHeader h;
+        u8 bus_id;
+        u8 bus_info;
+        u8 parent_bus;
+        u8 reserved[3];
+    };
+
+    struct [[gnu::packed]] CompatibilityBusAddressSpaceModifierEntry
+    {
+        ExtEntryHeader h;
+        u8 bus_id;
+        u8 address_modifier;
+        u32 predefined_range_list;
+    };
+
+}
+
+class PCIInterruptOverrideMetadata : public RefCounted<PCIInterruptOverrideMetadata> {
```

Comment:
This class doesn't seem like it needs reference counting. Is it expected to have multiple owners that keep it alive? Is it expected to be mutable?

Reply:
It's expected to have multiple owners - the Interrupt management will hold a reference and the corresponding IOAPIC object which handles the overriding.